M2354: Support Nuvoton's new target

ccli8 opened the pull request:
<!--
For more information on the requirements for pull requests, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html).

NOTE: Do not remove any of the template headings (even for optional sections) as this
template is automatically parsed. 
-->

### Summary of changes <!-- Required -->

M2354,is Nuvoton's CM23 based and TrustZone enabled chip. Based on [TF-M 1.3 release](https://git.trustedfirmware.org/TF-M/trusted-firmware-m.git), this PR enables TF-M 1.3 on Mbed OS 6.

**Mbed CLI build**:
```sh
mbed compile -m NU_M2354 -t ARM
```

**Mbed CLI2 build**:
```sh
mbed-tools compile -m NU_M2354 -t ARM
```
----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

<!--
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front of them would change the meaning incorrectly. 
-->
    [X] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

<!--
    Provide all the information required, listing all the testing performed. For new targets please attach full test results for all supported compilers.
-->
    [] No Tests required for this change (E.g docs only update)
    [] Covered by existing mbed-os tests (Greentea or Unittest)
    [x] Tests / results supplied as part of this PR
    
----------------------------------------------------------------------------------------------------------------

ccli8 committed:
M2354: Support TF-M

1.  Configure non-secure target name to NU_M2354 (targets/targets.json). No NU_M2354_NS alias
2.  Following template target, enable image signing and concatenating in post-build process
    (1) Add post-build script (tools/targets).
    (2) Enable TF-M custom build by centralize relevant stuff imported from TF-M (COMPONENT_TFM_S_FW).
3.  Add M2354Code.merge_secure into whitelist of uvision6 (tools/export/uvision/__init__.py).
4.  Add M2354 CMSIS pack database (tools/arm_pack_manager/index.json).
5.  Configure stdio baudrate to 115200 to match TF-M port (platform/mbed_lib.json).
6.  Define CMSIS_NVIC_VIRTUAL to override NVIC_SystemReset with TF-M version (cmsis_nvic_virtual.h).
7.  Override tfm_ns_interface_xxx(...) to enable NS secure call:
    (1) At pre-rtos stage
    (2) In SVC context
8. Implement secure function call with tfm_platform_ioctl(...).
9. Combine stddriver_secure.h/c and hal_secure.h/c into platform_extra_secure.h/c.
10. Fix peripheral base to non-secure (PeripheralNames.h) (TrustZone-unaware since Mbed OS 6.0).
11. Fix NU_PORT_BASE/NU_GET_GPIO_PIN_DATA/NU_SET_GPIO_PIN_DATA to non-secure (PinNamesCommon.h) (TrustZone-unaware since Mbed OS 6.0).
12. NSC convention for StdDriver sys/clk (both TF-M and Mbed must follow)
    (1) SYS_ResetModule
        Usage: Replaced with SYS_ResetModule_S on Mbed OS
        Action: Make it inaccessible from Mbed (neither source nor NSC). Provide SYS_ResetModule_S on Mbed via platform ioctl instead.
    (2) CLK_GetXxx
        Usage: Called in bpwm/i2s/qspi/sc/sdh and system_M2354 on Mbed OS
        Action: Make them inaccessible from Mbed (neither source nor NSC). Re-provide them on Mbed via platform ioctl instead.
13. Remove DISABLE/ENABLE macro definitions in BSP to avoid name conflict with other modules
14. Change to TMR4/5 from TMR2/3 for implementing us_ticker/lp_ticker because TMR2 is used for TF-M NSPE test
15. Support cmake

NOTE: Export(uvision6) doesn't support TF-M target. To enable it for partial compile on Keil, force below function to return true.
      is_target_supported(tools/export/uvision/__init__.py)

ccli8 committed:
M2354: Reduce output image size

1.  Enable mcuboot no padding option
2.  Support output image format Intel Hex

ccli8 committed:
M2354: Support TRNG as entropy source on TF-M

1.  Update TF-M Secure bin enabling TRNG as entropy source
2.  Replaced with above, remove TRNG HAL stuff on Mbed, including platform extra secure functions, cmake, etc.

ccli8 committed:
M2354: Refine TF-M/Mbed integration readme

Patater reviewed (commented):
@0xc0170 
I've reviewed the frozen tools files. The files modified will impact the online compiler (`tools/targets/__init__.py `, `tools/export/uvision/__init__.py`). However, the risk is low enough we can merge without a passing "Frozen tools check".

ccli8 committed:
M2354: Use TF-M exported region_defs.h to replace target.mbed_rom_start and friends

0xc0170 commented on targets/TARGET_NUVOTON/TARGET_M2354/TARGET_TFM/TARGET_NU_M2354/COMPONENT_TFM_S_FW/README.md:
```diff
@@ -0,0 +1,132 @@
+Navigate [TF-M](https://www.trustedfirmware.org/projects/tf-m/).
+Then go through **DOCS** → **Getting Started Guides** → **Software requirements** for TF-M build environment setup.
+
+**NOTE**: For unknown reason, **GNU Arm Embedded Toolchain 10-2020-q4-major** built code **FAILS** to run. Avoid this toolchain version.
```
is this reported anywhere?

ccli8 replied:
No. The cause is unknown. The TF-M is built on **GNU Arm Embedded Toolchain 9-2020-q2-update**.

LDong-Arm replied:
@0xc0170 Unfortunately we need to avoid 10-2020-q4-major, see https://review.trustedfirmware.org/c/TF-M/trusted-firmware-m/+/9284. It's fixed in the development version of GCC but we don't know when a newer release will come.
@ccli8 Does it looks related?

ccli8 replied:
Actually, we didn't get to the root cause 10-2020-q4-major failing. Just cross-check and do no gcc compiler upgrade to this version.

ccli8 committed:
M2354: Add serial_tx_empty(...)

Enable serial sync with the below PR:
https://github.com/ARMmbed/mbed-os/pull/14600

ccli8 committed:
M2354: Update to TF-M 1.3

LDong-Arm commented on targets/TARGET_NUVOTON/TARGET_M2354/TARGET_TFM/TARGET_NU_M2354/COMPONENT_TFM_S_FW/CMakeLists.txt:
```diff
@@ -0,0 +1,27 @@
+# Copyright (c) 2020-2021 ARM Limited. All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
+
+include(${MBED_PATH}/targets/TARGET_NUVOTON/scripts/mbed_set_post_build_nuvoton.cmake)
```
For Mbed CLI 2, we have provided a generic post build for all TF-M targets:
* [CMake definition](https://github.com/ARMmbed/mbed-os/blob/mbed-os-6.10.0/platform/FEATURE_EXPERIMENTAL_API/FEATURE_PSA/TARGET_TFM/TARGET_TFM_LATEST/scripts/mbed_set_post_build_tfm.cmake)
* [python script to run MCUboot imgtool](https://github.com/ARMmbed/mbed-os/blob/mbed-os-6.10.0/platform/FEATURE_EXPERIMENTAL_API/FEATURE_PSA/TARGET_TFM/TARGET_TFM_LATEST/scripts/generate_mbed_image.py)

To use it (e.g. Musca S1), simply do: https://github.com/ARMmbed/mbed-os/blob/9738b27c7df897c29e9769911d6794ba3e5b3f19/targets/TARGET_ARM_SSG/TARGET_MUSCA_S1/CMakeLists.txt#L4-L13

Is it possible to use it for M2354 too rather than having a dedicated one for this target?

ccli8 replied:
On M2354, we use a dedicated post-build script because:
1. The generic script supports only two-image signing (signing tf-m_s/mbed separately), but M2354 uses one-image signing (signing tf-m/mbed together).
1. In M2354 dedicated script, during signing, the padding option `--pad` is removed (so disable upgradeable by mcuboot) to decrease binary size for drag-n-drop flash. The having `--pad` version is still generated to enable future upgrade by mcuboot.
1. Difference of memory layout with ITS, NV counter, etc. can cause the generic script to not fit M2354.

LDong-Arm replied:
Thanks for the explanation. A separate script is fine now.

Maybe in the future we (Mbed OS) could consider making the generic script a bit more versatile - it should be quite doable given that the trusted-firmware-m project has a unified signing mechanism for all targets if I understand correctly. @ARMmbed/mbed-os-core FYI.

LDong-Arm reviewed (commented):
Thanks @ccli8, it looks good to me.

Because of the frozen tools check failure in Travis, tests that come after it were not run. Given the scope of this PR, we should have the other checks run and passing. I've created #14665, once it's merged could you please rebase this PR?

@ARMmbed/mbed-os-hal Do you have any comments on the target support and HAL implementation?

MarceloSalazar commented on targets/TARGET_NUVOTON/TARGET_M2354/TARGET_NU_M2354/PinNames.h:
```diff
@@ -0,0 +1,115 @@
+    PG_2    = NU_PORT_N_PIN_TO_PINNAME(6, 2), PG_3, PG_4, PG_9 = PG_4 + 5, PG_10, PG_11, PG_12, PG_13, PG_14, PG_15,
+    PH_4    = NU_PORT_N_PIN_TO_PINNAME(7, 4), PH_5, PH_6, PH_7, PH_8, PH_9, PH_10, PH_11,
+
+    // Arduino UNO naming
```
Please follow the guidelines for Arduino Uno connector
https://os.mbed.com/docs/mbed-os/v6.10/porting/standard-pin-names.html

ccli8 replied:
Fixed

MarceloSalazar commented on targets/TARGET_NUVOTON/TARGET_M2354/TARGET_NU_M2354/PinNames.h:
```diff
@@ -0,0 +1,115 @@
+#if defined(MBED_CONF_TARGET_USB_UART_TX)
+    CONSOLE_TX      = MBED_CONF_TARGET_USB_UART_TX,
+#else
+    CONSOLE_TX      = NC,
```
A pin cannot be defined as NC

ccli8 replied:
Fixed

MarceloSalazar commented on targets/TARGET_NUVOTON/TARGET_M2354/TARGET_NU_M2354/PinNames.h:
```diff
@@ -0,0 +1,115 @@
+
+    // Button naming
+    SW2 = PF_11,
+    BUTTON1 = SW2,
```
Please follow the guidelines for LEDs/BUTTONs
https://os.mbed.com/docs/mbed-os/v6.10/porting/standard-pin-names.html

ccli8 replied:
Fixed

MarceloSalazar reviewed (dismissed):
Please fix pin-names

ccli8 committed:
M2354: Comply with standard pin naming rule

MarceloSalazar reviewed (dismissed):
LGTM

LDong-Arm reviewed (approved):
LGTM
